### ISSUE TITLE
Non-optimal distribution of training data

### ISSUE THREAD

**Issue by mooskagh** [glinscott/leela-chess#513]
As discussed in discord,

Currently we pick every move uniformly from every game, which makes 1% of training chunks to be starting position, and another 1% of chunks is position after 1. e4., probably near another 1% after 1. e4 e5, and in general I expect that something like 10% of training data contains less than a hundred of unique positions (speculative guess).

That probably makes it overfit in opening.

I think it makes sense to make the distribution more diverse by sampling out openings more agressively.

**Comment by jkiliani** [comment]
I'm actually not sure this reasoning is correct, because while the positions are identical (and the policy targets as well, barring the influence of noise), game outcomes are not. The current procedure trains a lot on those positions, true, but doesn't it have to in order to have the value head converge correctly for the opening? If we did sample out opening positions during training, we would have to be extremely careful that the distribution of game results for identical positions did not change with the sampling.

**Comment by mooskagh** [comment]
For value head maybe, but for policy head I guess they basically all feed the same data.

**Comment by jkiliani** [comment]
Yes true, but is there any harm in that? A well trained value head is highly important for playing strength, especially scaling with visits...

**Comment by jjoshua2** [comment]
I think if opening isn't distributed enough, the best way to fix that is for the server to assign a random opening from the SF 2 move book. It will save 4 ply of work from each client, and make sure maximum variety of opening is used in training so there is less duplication of work and training data. Probably have 10% still play from start would be a wise move to not lose opening knowledge from the first 2 moves.

**Comment by jkiliani** [comment]
I think one possible option here would be to collate the training data of identical positions somehow, by averaging both visit distribution and value training targets for those cases and then training less on these averaged training positions. It would take quite a bit of effort though to do this, but it may actually make the training procedure more stable for the opening. Instead of training on 1. e4 100,000 times with different game results each time, we could train much less steps on this position, but each time train the value head to 0.515 (or whatever averaging game results will yield).

**Comment by patrickchess** [comment]
A possible solution to "overweighting" the starting position and first move (e4) may be to sample moves non-uniformly. Most crudely, you could reduce the probability of selecting the first few moves by a factor of 2 or more.

**Comment by Error323** [comment]
After discussing with @mooskagh we think that it's most optimal to hash every position into a leveldb database and read sequentially from there during training. This makes sure every position exists exactly once per training window. 

So before a training cycle, unpack all chunks sequentially, hash (with a random seed per training cycle) on position state and history. Then feed them in level db. Initiate training and read the [leveldb](https://github.com/google/leveldb#write-performance).

**Comment by jkiliani** [comment]
@Error323 If you make positions unique per training window, how do you choose your training targets? Are you averaging over the game results and visits counts (which are not identical either because of Dirichlet noise)?

Doing so will significantly change the MSE loss landscape by the way, but it may well be a positive change...

**Comment by grolich** [comment]
It does seem a bit risky, as in general, positions which occur a lot more are a lot more useful for the net to learn than positions which are very rare.

Perhaps you are just referring to using some different weighting mechanism for pulling info from the positions db to still give a higher chance to positions which occur more frequently but not letting it get too extreme? if so, I would imagine some testing would be needed to make sure what strategy would be more effective and/or risky.
